### PR TITLE
Centralise exception handling of `MultiplayerClient` calls

### PR DESCRIPTION
**Pull request by peppy**
Feels very flaky that we are firing async methods into the wind. This is a first step at bringing some sanity to the usage of `MultiplayerClient`.

I'm sure I've missed some calls, but more interested in ensuring that everyone is on board with the change.

### CODE REVIEW

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs [hunk 1]
```diff
@@ -76,40 +72,23 @@ protected override void SelectItem(PlaylistItem item)
...
-                                // We generally use the message field for a user-parseable error (eventually to be replaced), so drop the first part for now.
-                                ? exception.Message.Substring(exception.Message.IndexOf(':') + 1).Trim()
-                                : exception.Message;
-
-                            Logger.Log(message, level: LogLevel.Important);
-                            Carousel.AllowSelection = true;
-                            return;
-                        }
+                    loadingLayer.Hide();
+                    this.Exit();
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs:82]: Given that the `onSuccess` and `onError` paths here are quite similar, maybe worth splitting a helper? Not sure.

**Review by bdach — COMMENTED** [review]
Code that is here looks mostly sane, although I haven't gone through multiplayer with a fine toothed comb to check if any calls were missed. Primarily interested in confirming the direction after the previous attempt of adding something like this (see above).

**Comment by peppy** [comment]
> Getting shades of "time is a flat circle" here again, since a similar sort of thing was already added in #11266 and then removed again in #11622 in favour of changing framework to not die on unobserved exceptions...
> 
> Has something changed in this respect that this has been brought back? I don't really have any reason to object to either solution, I just don't want to go back and forth in perpetuity.

I had strong feelings that I was doing this not for the first time....

I'm not sure why the first one was removed, but I really think that `async` calls should never be *intentionally* fired and forgotten. Decorating these methods at least makes it obvious that the method calls are `async` execution... since they don't even have a suffix right now.

Unobserved handling should be the last catch that prevents the game from crashing. I guess last time around the local handling was removed because there was "no reason to leave it in place" but I disagree with my decision to remove it now.

**Comment by smoogipoo** [comment]
Unobserved exceptions don't cause the application to crash. One thing that's gained from this is more timely and better exception logging.

**Comment by peppy** [comment]
> Unobserved exceptions don't cause the application to crash. One thing that's gained from this is more timely and better exception logging.

They did before https://github.com/ppy/osu-framework/pull/4171. My reply was responding to @bdach who touched on the change where I made the framework *not* crash, then reverted a similar change to the one made in this very PR.

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs:82]: A local method local to this class you mean? Something like

```csharp
FireAndForget(() => handle(null), handle)

handle(Exception e)
{
...
}
```

?

Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs:82]: Yeah, just something local to remove the duplication of the current screen check and hiding the loading layer. Although I'm not entirely 100% it's worth the effort.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs]: Not sure it's worth it. I've refactored the code a bit, as the `IsCurrentScreen` check wasn't required for most of it and was just a safety.

**Comment by smoogipoo** [comment]
Tests need some attention

**Review by bdach — APPROVED** [review]
seems ok

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**